### PR TITLE
[Sprite Lab Animations] Animation Upload Refactor and new size requirements

### PR DESCRIPTION
**Pull request by mikeharv**
This is complete work of what was started here:
- https://github.com/code-dot-org/code-dot-org/pull/47913

Code Review revealed some missed edges that could lead the component to an incorrect state. It ended up being easier to essentially start from scratch on a new branch.

### New Size Requirements
All animations must be between 100x100 pixels* (the default rendered size of a sprite in Sprite Lab), and 800x800 (double the dimensions of the Sprite Lab canvas, allowing for large high resolution sprites). Users **cannot** upload images outside this range.
* Technically, images only need to be 100px in either height or width. For example, a 75x100 image is okay because a sprite using it for an animation would still look fine at the default size.

_Note: The previous PR set this limit as 970x970. This was limit was based on an edge case where we were incorrectly calculating the dimensions of super-large files, based on their resized preview. We now correctly calculate height and width for all images._

### ValidationStep refactor
@epeach pointed out that we have a nice component that already exists for showing the status of various checks. Today it is used in our Maker Setup context as well. This component was used here in several places.

### Events and State
Special attention was paid to make sure that the component state was always updated and/or reset as a result of each event. If a "later" part of the state depends on "earlier" information, we clear that out if the user makes any adjustments within the earlier steps. We can also reset much of the component state when an upload completes successfully. We also track state for manually entered information, such as availability, categories, and aliases. These state properties do not need to be reset. 

<details><summary><h3>Click to expand for screenshots of the current versions on Levelbuilder</h3></summary>

## Before (On page load)
imageimage

## Before (Completed)
imageimage

</details>

## Screenshots
Each set of screenshots below shows the initial view on page load, followed by an expected view after use and some other examples, including expected errors.

### Step 1
step 1 initial
step 1 passing
Step 1 errors
step 1 errors2

### Step 2
step 2 initial
step 2 passing
To prevent users from interacting with the radio buttons before we have information about the existing animations, we hide this with a separate ValidationStep message until everything has loaded:
step 2 waiting
Below: If a user has already generated metadata and then changes availability or category, they must re-generate the metadata. In order to do so, they must either choose "level-specific animation" OR choose "library animation" WITH a category.
step 2 change
Whenever the user interacts with the radio buttons or dropdown we check to see if there is already an animation with that path and name:
step 2 error

### Step 3
step 3 initial
The button is only enabled if there is metadata (and the other requirements before it):
image
image
For this screenshot, I turned off my WiFi right before clicking the upload button:
step 3 error

### Edge Case Testing
Beyond testing typical use of the feature, I also looked to make sure the state and what was rendered was always accurate based on how the user interacted with the page. For example:

- Clicking "Choose File", but then canceling before selecting an image.
- Dragging a non-png file over the "Choose File" button, instead of clicking it.
- Adding files with different extensions, dimensions, and filename structures.
- Adding files with names that match those in the existing libraries.
- Setting the availability/category/aliases before choosing/changing the file.
- Changing the availability/category/aliases/file after generating the metadata.

There is one known "bug" that I can think of where one would be allowed to upload a animation with a name that already exists if they had just uploaded the first one during this same session and page load. For example, if a user uploads `new_cat` and then immediately afterwards, without page load, uploads a different file that is also called `new_cat`. Unless we re-retrieve the list of all uploaded animations with each upload (again, VERY slow), we don’t know that the newly uploaded animation is there until refreshing the page. I tried forcing that re-retrieval and it made for a painfully slow experience, so I'd recommend we leave this as is. (I also think it unlikely that a curriculum writer would accidentally upload two files with the same name during the same session, and if they did it’s probably a case where they actually do want to override it - because they’ve tweaked it or something.)

## Links

- jira ticket: [STAR-2379: [Sprite Upload] UI - Add a UI warning when uploaded image is larger than expected](https://codedotorg.atlassian.net/browse/STAR-2379)

## Follow-up work

It should be noted that it still takes a very long time (average 80 seconds) to retrieve the full list of level animations. If we can make this more efficient it would improve the user experience.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/code-studio/assets/SpriteLabAnimationUpload.jsx
Comment (epeach) [inline apps/src/code-studio/assets/SpriteLabAnimationUpload.jsx]: I like renaming this to align with our sprite -> animation naming decisions, but I'm hesitant about renaming the file to SpriteLabAnimationUpload because using SpriteLab specific variable naming. Most (if not all) of this component can be reused for GameLab animation management in the future. I'd love to leave that path open for an extension of this work. Happy to chat a bit about feasibility and renaming!

Comment (mikeharv) [reply apps/src/code-studio/assets/SpriteLabAnimationUpload.jsx]: How about just AnimationUpload.jsx?

Comment (epeach) [reply apps/src/code-studio/assets/SpriteLabAnimationUpload.jsx]: :+1:

File: apps/src/code-studio/assets/SpriteLabAnimationUpload.jsx
Comment (epeach) [inline apps/src/code-studio/assets/SpriteLabAnimationUpload.jsx]: Because of the delay here, we might want to considering adding 'loading' spinner? Not sure if that would improve the UX or draw more attention to the delay?

Comment (epeach) [reply apps/src/code-studio/assets/SpriteLabAnimationUpload.jsx]: Ahhh, whoops, I see you already did that! Good call!

File: apps/src/code-studio/assets/SpriteLabAnimationUpload.jsx
Comment (epeach) [inline apps/src/code-studio/assets/SpriteLabAnimationUpload.jsx]: Do we want to include 'rename and reupload'? I think that clarifies that the extra action here is to edit the file name locally and then re-start this upload process.

File: apps/src/code-studio/assets/SpriteLabAnimationUpload.jsx
Comment (epeach) [inline apps/src/code-studio/assets/SpriteLabAnimationUpload.jsx]: As we are renaming the file and doing some deeper refactor work here, I think this is a good chance for us to pause and break out this mega component into smaller components. That'll make future work (and reviews!) way easier. Thoughts?

Comment (mikeharv) [reply apps/src/code-studio/assets/SpriteLabAnimationUpload.jsx]: 100% agree!! My plan for now was to get this huge refactor all working (in one place), and then divide it out into smaller components as a next step. I'd love to pick your brain on the best way to do this. I think it's something that could happen on this branch before anything gets merged.

Comment (mikeharv) [reply apps/src/code-studio/assets/SpriteLabAnimationUpload.jsx]: @epeach Actually on second thought, I do wonder about possibly making this an "up next" refactor task. I know that curriculum is getting ready to upload more final HoC assets and it would be cool to have them use this new version if that were possible.

File: apps/src/code-studio/assets/SpriteLabAnimationUpload.jsx
Comment (mikeharv) [reply apps/src/code-studio/assets/SpriteLabAnimationUpload.jsx]: Happy to add that. I kept it short to avoid needing to wrap to a second line in the UI, but I don't think there's a real reason we need to do that, especially at the cost of clarity.

File: apps/src/code-studio/assets/SpriteLabAnimationUpload.jsx
Comment (mikeharv) [reply apps/src/code-studio/assets/SpriteLabAnimationUpload.jsx]: Reverted to SpriteUpload.jsx for this PR. As discussed, we'll handle any renaming of files when we do the follow-up work to break this into separate components.

File: apps/src/code-studio/assets/SpriteUpload.jsx
Comment (epeach) [inline apps/src/code-studio/assets/SpriteUpload.jsx:12]: Not sure, but I wonder about storing these as an object or enum. I don't think that buys us anything programmaticly, but it migth help organize these values as "Errors". No action needed, but wanted to put a bug in your ear.

File: apps/src/code-studio/assets/SpriteUpload.jsx
Comment (epeach) [inline apps/src/code-studio/assets/SpriteUpload.jsx:43]: I like this chunking of state. I know it's an intermediate phase until we start to break this file out into smaller components, but I think this is a good organization strategy for right now.

File: apps/src/code-studio/assets/SpriteUpload.jsx
Comment (epeach) [inline apps/src/code-studio/assets/SpriteUpload.jsx:84]: A more descriptive name (or comment) might help readability for what exactly is being submitted.

File: apps/src/code-studio/assets/SpriteUpload.jsx
Comment (epeach) [reply apps/src/code-studio/assets/SpriteUpload.jsx:12]: Okay, looking onward at how this is issued - I think I do prefer this as is. :)

**Review by epeach — APPROVED** [review]
Some small nits, but nothing to hold off for. In general, I think a change this big might benefit from smaller, more incremental changes to make it easier to conceptualize and compartmentalize all the changes. Overall though, thank you for taking on these awesome improvements!

**Pull request merged**
